src/mcp/server/session.py

```diff
@@ -121,27 +133,51 @@ def check_client_capability(self, capability: types.ClientCapabilities) -> bool:
         if capability.sampling is not None:
             if client_caps.sampling is None:
                 return False
-            if capability.sampling.context is not None:
-                if client_caps.sampling.context is None:
-                    return False
-            if capability.sampling.tools is not None:
-                if client_caps.sampling.tools is None:
-                    return False
-
-        if capability.elicitation is not None:
-            if client_caps.elicitation is None:
+            if capability.sampling.context is not None and client_caps.sampling.context is None:
+                return False
+            if capability.sampling.tools is not None and client_caps.sampling.tools is None:
                 return False
 
+        if capability.elicitation is not None and client_caps.elicitation is None:
+            return False
+
         if capability.experimental is not None:
             if client_caps.experimental is None:
                 return False
-            # Check each experimental capability
             for exp_key, exp_value in capability.experimental.items():
                 if exp_key not in client_caps.experimental or client_caps.experimental[exp_key] != exp_value:
                     return False
 
+        if capability.tasks is not None:
+            if client_caps.tasks is None:
+                return False
+            if not check_tasks_capability(capability.tasks, client_caps.tasks):
+                return False
+
         return True
 
+    def set_task_result_handler(self, handler: ResponseRouter) -> None:
```

Comment:
all this does is wrap another function, do we need this? can't we just `self.add_response_router` directly somewhere